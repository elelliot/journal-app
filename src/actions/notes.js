import { db } from "../firebase/firebase-config";
import { collection, addDoc, updateDoc, doc, deleteDoc } from "firebase/firestore";

import { types } from '../types/types';
import { loadNotes } from "../helpers/loadNotes";
import Swal from 'sweetalert2';
import { fileUpload } from "../helpers/fileUpload";

//Estructura de la db de las notas: uid (user) -> journal -> notes -> id( autogenerado) de la nota -> nota


export const startNewNote = () => {

    return async ( dispatch, getState ) => {

        //obtenemos el State, literal todo, y con esto podemos obtener cualquier cosa del estado
        // const state = getState();
        // console.log(state);

        //obtenemos uid del user (obtenido desde el state) 
        const { uid } = getState().auth;
        
        //nueva nota
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime() //momento exacto en que se crea la nota
        }

        //Query, no es obligatorio, pero es buena práctica hacerlo con try catch
        //agregamos a la db, la nueva nota con id autogenerado en la colección dada (si no existe, la crea).

        let docRef = '';
        try {
            docRef = await addDoc(collection(db, `${uid}/journal/notes`), newNote)
            //console.log("New Note Added",docRef);
          } catch (e) {
            console.log("Error: Couldn't create Note",e)
          }


        //mandamos al reducer
        dispatch( activeNote( docRef.id, newNote ) );

        //Refresh Sidebar
        dispatch( addNewNote(docRef.id, newNote) );
    }
}

//ya creamos la nota, ahora la queremos visualizar al crearla
export const activeNote = ( id, note ) => ({
    type: types.notesActive,
    payload: {
        id, //id de la nota
        ...note //el resto igual con todo y id
    }
});

//para que al crear la nota, haga refresh el sidebar
export const addNewNote = ( id,note ) => ({
    type: types.notesAddNew,
    payload: {
        id, ...note
    }
})

export const startLoadingNotes = ( uid ) => {
    return async( dispatch ) => {
        //loadNotes es una promesa..., y después 
        const notes = await loadNotes( uid );
        //aquí es cuando se mandan a Redux
        dispatch( setNotes( notes ) );
    }
}
export const setNotes = ( notes ) => ({
    type: types.notesLoad,
    payload: notes //como siempre estamos creando un nuevo arreglo, no necesitamos el spread ya que sería "redundante"
});


//Update en la BD al dar click en Save
export const startSaveNote = ( note ) => {
    return async( dispatch, getState ) => {

        //obtenemos uid del user (obtenido desde el state) 
        const { uid } = getState().auth;

        //Si no viene el url de la imagen, lo borramos para evitar meter un undefined a firebase ya que dará error
        if( !note.url ) {
            delete note.url;
        }


        //El id de la nota no lo queremos grabar en FireStore, por eso lo eliminamos en nu nuevo obj.
        const noteToFireStore = { ...note };
        delete noteToFireStore.id;

        //Update
        const noteRef = doc(db, `${uid}/journal/notes/${note.id}`);
        try {
            await updateDoc(noteRef, noteToFireStore);
            //console.log('Note Updated!');
            
            //de esta forma el sideBar se actualiza al oprimir save, el problema es que además de la nota que cambió, también se actualizan las demás
            //dispatch(startLoadingNotes( uid );
            dispatch( refreshNote( note.id, noteToFireStore) );

            //Saved Note Alert
            Swal.fire({
                iconColor: '#41D70C',
                icon: 'success',
                titleText: `Note: "${ noteToFireStore.title }" Saved!`,
                color: "#FFFFFF",
                background: "#4D4D4D",
              })


          } catch (e) {
            console.log('There was an error while updating note!',e);
          }
    }

}

//
export const refreshNote = ( id, note ) => ({
    type: types.notesUpdate,
    payload: {
        id,
        note: {
            id,
            ...note //Recordar que a noteToFirestore le borramos el id  
        }
    }
})


//subimos la imagen (es asíncrona)
export const startUploading = ( file ) => {
    return async( dispatch, getState ) => {
        
        //obtenemos la activeNote y la renombramos
        const { active: activeNote } = getState().notes;

        //Para que se muestre "cargando" al subir imagen...
        Swal.fire({
            title: 'Uploading...',
            text: 'Please wait...',
            allowOutsideClick: false,
            showConfirmButton: false,
            willOpen: () => {
                Swal.showLoading();
            }
        });
        
        //Subimos el file a Cloudinary y obtenemos el url de la imagen subida a Cloudinary
        const fileUrl = await fileUpload( file );
        //console.log(fileUrl);

        //Le asignamos el url de la imagen a la nota activa y lo mandamos con el dispatch a FireStore
        activeNote.url = fileUrl;
        await dispatch( startSaveNote( activeNote ) );

        Swal.close();//quitamos la carga en cuanto la imagen se haya subido a Cloudinary
    }
}


//Borrar la nota de FireStore
export const startDeleting = ( id ) => {

    return async( dispatch, getState ) => {

        //obtenemos uid de la nota actual
        const { uid } = getState().auth;

        //referencia a la nota activa(id) del usuario (uid)
        const noteRef = doc(db, `${uid}/journal/notes/${id}`);
        await deleteDoc(noteRef); //borramos la nota
        
        //Modificamos el Store
        dispatch( deleteNote(id) );
    }
}

//Borramos la nota del Store
export const deleteNote = ( id ) => ({
    type: types.notesDelete,
    payload: id
});


//Limpiar las notas del state al dar logout
export const noteLogout = () => ({
    type: types.notesLogoutCleaning
});