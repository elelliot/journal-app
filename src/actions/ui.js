import { Toast } from "../alerts/alert";
import { types } from "../types/types";

//No necesitamos que sean async actions
// export const setError = ( err ) => ({
//     type: types.uiSetError,
//     payload: err
// });

export const setError = (err) =>{
    Toast.fire({
        icon: 'error',
        title: err
      });
      return{
        type: types.uiSetError,
        payload: err
      }
}


export const removeError = () => ({
    type: types.uiRemoveError
});

//Controlamos el estado del loading mandando el type de accion
export const startLoading = () => ({
    type: types.uiStartLoading
});


export const finishLoading = () => ({
    type: types.uiFinishLoading
});