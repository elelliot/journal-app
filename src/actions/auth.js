

import { getAuth, signInWithPopup, createUserWithEmailAndPassword, signInWithEmailAndPassword, updateProfile, signOut } from "firebase/auth";
import { Toast } from "../alerts/alert";
import { googleAuthProvider } from "../firebase/firebase-config";
import { types } from "../types/types"
import { noteLogout } from "./notes";
import { finishLoading, startLoading } from "./ui";

//Alerts ver en alert.js
const welcome = () => { Toast.fire({
    iconColor: '#41D70C',
    icon: 'success',
    title: 'Welcome, what do we have for today? ^u^/'
  })};

const loginErr = () => {Toast.fire({
    icon: 'error',
    title: 'User or Password might be wrong or don\'t exist, try again ^u^'
  })};


const emailUsed = () => {Toast.fire({
    icon: 'error',
    title: 'Email already used'
  })};
//actions


//acción que dispara otra acción(login) cuando el setTimeout(tarea asíncrona) se resuelve
export const startLoginEmailPassword = (email, password) => {

    //al usar el dispatch de redux, la funcion retorna un callback ()=> ,y uno de sus args es el dispatch de Thunk  (todo adentro del callback se ejecuta)
    //se espera que se use el dispatch para trabajar los datos que llegan de una tarea async

    //se pueden realizar fetch, posteo de archivos o incluso dispatch de otra acción dentro del callback
    return (dispatch) => {
    
        //loading: true
        dispatch( startLoading() );

        //Para loguearse con email y password
        const auth = getAuth();
           return signInWithEmailAndPassword(auth, email, password)
            .then( ({user}) => {
                //console.log(user)
                //Mandamos uid y displayName al state cuando se hace login
                //dispatch( login( user.uid, user.displayName) );
                
                //PA LAS PRUEBAS puse el return y el dispatch del login(con esto se dispararía 2 veces en la app):
                dispatch( login( user.uid, user.displayName) );
                
                welcome()
            })
            .catch( ({ message }) => {
                
                //Args: Title, HTML, Icon... Alerta de error de Sweet Alert2
                //Swal.fire('Error', message, 'error');
                
                loginErr()
            })
            .finally( () => {
                //Para no tener que usar el finishLoading 2 veces podemos usarlo una vez aquí y en cualquiera de los 2 casos se llama
                //loading: false
                dispatch( finishLoading() );
            });
    }
}



//Registrar email
export const startRegisterWithEmail = ( email, password, name ) => {
    //en cuanto tenga el user grabado en firebase realizaré el dispatch
    return( dispatch ) => {

        const auth = getAuth();
        //en cuanto se crea el user esto intenta autenticar inmediatamente por eso da error por ahora.
        createUserWithEmailAndPassword(auth, email, password)
            .then( async ({ user }) => {
                
                //"displayName" es null usando esta función, solo se obtiene cuando se usa una red social como login
                //podemos asignar el name(el que damos en el form) al displayName con updateProfile
                await updateProfile(user,{displayName: name})
                
                //ya que el useEffect se ejecuta antes del updateProfile, el name es null, así que por ahora haremos el mismo dispatch con el nombre obtenido del update y tendremos 2 login al terminar el registro
                dispatch( login(user.uid, user.displayName ) )
                welcome()
            })
            .catch( ({ code } ) => {
                //desestructuramos message de error
                
                emailUsed()
            });
    }
}



//login con google 
export const startGoogleLogin = () =>{
    return (dispatch) =>{
        const auth = getAuth();
        //(aquí me tira el popup de login)
        signInWithPopup(auth, googleAuthProvider)
        //se obtiene el parametro por default y desestructuramos user del mismo
            .then( ({ user }) =>{
                //desestructuramos para obtener uid y displayName (user es solo una propiedad del objeto grande que te da la promise al cumplirse, por eso se ponen así en específico):
                //dispatch( login( user.uid, user.displayName ) );
                
                welcome()
            })
            .catch( ( { code } )  => {
                //por si se cierra el popup sin haber seleccionado usuario
                
                // const errorCode = error.code;
                // const errorMessage = error.message;
                //console.log(message);
                Toast.fire({
                    icon: 'error',
                    title: code
                  })
            });
    }
}


//El login que se usa en las otras 2 acciones
export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid, 
        displayName
    }
});


//Logout es Async por que hay que se usa una función firebase para esto y usa promesa

export const startLogout = () => {
    return async( dispatch ) => {
        const auth = getAuth();
        await signOut(auth)
            .then( ()=>{

                //limpiamos el user del state 
                dispatch( logout() );
                //limpiamos las notas del state
                dispatch( noteLogout() );
            })
            .catch( (e) => {
                console.log(e); //no suele pasar pero en caso de que se vaya el internet o algo
            })
    }
}

export const logout = () => ({
    type: types.logout,
});


//forma larga del return
// export const login = (uid, displayName) => {
//     return {
//         type: types.login,
//         payload: {
//             uid, 
//             displayName
//         }
//     }
// }


