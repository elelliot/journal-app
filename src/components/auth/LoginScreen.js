import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Link } from 'react-router-dom'

import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth'
import { useForm } from '../../hooks/useForm'

export const LoginScreen = () => {
    
    //igual que en el RegisterScreen, obtenemos el loading del ui del state de redux
    //esto va en el disabled = del botón de login
    const { loading } = useSelector( state => state.ui );
    
    //dispatch de acciones al store con redux 
    const dispatch = useDispatch();

    //Form
    const [ formValues, handleInputChange ] = useForm({
        email: 'elelliot@gmail.com',
        password: '123456'

    });

    const { email, password } = formValues;



    //Submit
    const handleLogin = (e) => {
        e.preventDefault();//para que no se refresque
        // console.log(email, password)
        //dispatch(login(123,'ElElliot'))

        //dispatch(el de Redux) de acción al store, recibe email y password en el actions/auth.js
        dispatch( startLoginEmailPassword(email, password) );

    }

    //botonazo de google
    const handleGoogleLogin = () =>{

        //como usamos signInwithPopup, no evitamos el refresh, pero para movil, sería apropiado manejarlo
        dispatch( startGoogleLogin() )
    }

    return (
        // Formulario V1
        <div className='animate__animated animate__fadeIn animate__faster'>
            <h3 className="auth__title">Login</h3>
            <form onSubmit ={ handleLogin }>
                <input
                    type="text"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    autoComplete="on"
                    value = { email }
                    onChange = { handleInputChange }
                />

                <input
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input"
                    autoComplete="on"
                    value = { password }
                    onChange = { handleInputChange }
                />

                <button
                    type="submit"
                    className="btn btn-primary btn-block"
                    disabled={ loading }
                >
                    Login
                </button>

                {/* Botón de Google, se supone que el auth te dice que está en ese archivo, y luego iría la clase, pero something´s off*/}
                <div className="auth__social-networks">
                    <p>Login with social network</p>
                    <div 
                        className="google-btn"
                        onClick={ handleGoogleLogin }
                        >
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>

                <Link 
                    to="/auth/register"
                    className="link"
                >
                    Create new Account
                </Link>


            </form>
        </div>

    )
}
