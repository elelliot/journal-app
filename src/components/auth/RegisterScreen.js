import React from 'react';
import validator from 'validator';
import { useDispatch, useSelector } from 'react-redux';

import { Link } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { removeError, setError } from '../../actions/ui';
import { startRegisterWithEmail } from '../../actions/auth';

export const RegisterScreen = () => {
    //dispatch de acciones al store con redux
    const dispatch = useDispatch();

    /*useSelector es un hook que dispara un callback en el cual tengo el state completo del store de Redux
    queremos agarrar el mensaje de error, como está en el ui lo extraemos con su nombre.*/
    const {msgError} = useSelector( state => state.ui );

    //Recordar que React redibuja y dispara el mensaje con cualquier cambio, y el msgError por defecto es null
    //console.log(msgError);


    //Form
    const [ formValues, handleInputChange ] = useForm({
        name:'Elliot',
        email:'elelliot@gmail.com',
        password: 123456,
        password2: 123456
    });
    const { name, email, password, password2 } = formValues;
    
    //Submit
    const handleRegister = (e) =>{
        //Evitamos Refresh y propagación por el URL
        e.preventDefault();
        
        if ( isFormValid() ){
            
            //console.log('Formulario Correcto');
            //dispatch para registrar a firebase con correo (debe ser diferente al de google)

            dispatch( startRegisterWithEmail(email, password, name) );
        }

    }

    /*
    Validamos Formulario (en lugar de hacer RegEx una alternatva sería usar la librería de npm Validator)
    el <input type="email"> no se usó, why?
    -dispatch al store de cada mensaje de error
    */
    const isFormValid = () => {
        
        if( name.trim().length === 0){
            //Si no hay nada escrito en el nombre
            dispatch( setError('Name is Required') );
            
            return false;
        } else if( !validator.isEmail( email ) ){
            //si el email no es válido (usamos la librería de validator para eso)
            dispatch( setError('Email not Valid') );
            
            //console.log('Email not Valid'); 

            return false;
        } else if ( password !== password2 || password.length < 5 ) {
            //Si el Password no coincide o si es menor a 5 car. 
            //QUEDA PENDIENTE CAMBIAR EL COMPORTAMIENTO (VER MI PREGUNTA EN UDEMY CLASE 251)
            dispatch( setError('Password should be at least 6 characters and Match each other') );
            
            //console.log('Password should be at least 6 characters and Match each other');
            return false;
        }
        //si es válido
        dispatch( removeError() ); 
        return true;
    }


    return (
        <div className='animate__animated animate__fadeIn animate__faster'>
            <h3 className="auth__title">Register</h3>

            
            {
                //Mensaje de Error del Formulario se maneja en la acción setError para usar Sweet Alert, 
                //ya que el msgError es null cuando queremos usarlo en el isFormValid y usarlo aquí rompe la app
                //si msgError es !== null muestra el mensaje de error
                msgError &&
                (
                    <div className="auth__alert-error">
                        { msgError }
                    </div>
                )
            }

            <form onSubmit ={ handleRegister }>
            <input
                    type="text"
                    placeholder="Name"
                    name="name"
                    className="auth__input"
                    autoComplete="off"
                    value={ name }
                    onChange = { handleInputChange }
                />
                
                <input
                    type="text"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    autoComplete="off"
                    value={ email }
                    onChange = { handleInputChange }
                />

                <input
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input"
                    autoComplete="on"
                    value={ password }
                    onChange = { handleInputChange }
                />

                <input
                    type="password"
                    placeholder="Confirm Password"
                    name="password2"
                    className="auth__input"
                    autoComplete="on"
                    value={ password2 }
                    onChange = { handleInputChange }
                />


                <button
                    type="submit"
                    className="btn btn-primary btn-block mb-5"
                    // disabled={ true }
                >
                    Register
                </button>

                <Link 
                    to="/auth/login"
                    className="link "
                >
                    Already Registered?
                </Link>


            </form>
        </div>
    )
}
