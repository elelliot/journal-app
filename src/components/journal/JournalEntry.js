import React from 'react';
import moment from 'moment';
import { useDispatch } from 'react-redux';
import { activeNote } from '../../actions/notes';


export const JournalEntry = ({ id, date, title, body, url}) => {
    
    //para manejar las fechas de los entries
    const noteDate = moment(date);
    //console.log(noteDate);

    const dispatch = useDispatch();
    const note = {
        date,
        title,
        body,
        url
    }

    //al hacer click activamos la nota (Faltaría cargar la info, solo es para Redux)
    const handleEntryClick = () => {
        dispatch( activeNote(id, note) );
    }

    return (
        //cada entrada renderizada básada en la cantidad del array entries del JournalEntries.js
        <div className="journal__entry pointer animate__animated animate__fadeIn animate__faster" onClick={ handleEntryClick }>
            {/* Foto del entry */}
            {
                //Si el url de la imagen existe, muestra la imagen, de lo contrario, no muestra nada
                url && 
                <div 
                className="journal__entry-picture"
                //Los estilos en react se ponen como objeto de javascript por eso las 2 llaves, es expresión y objeto
                //si el estilo tiene nombre con guión, se usa camelCase
                style={{
                    backgroundSize:'cover',
                    backgroundImage:`url(${ url })`
                }}
                ></div>
            }

            {/* Cuerpo (Contenedor), titulo y contenido de cada entry */}
            <div className="journal__entry-body">
                <p className="journal__entry-title">
                  { title }  
                </p>
                <p className="journal__entry-content">
                    { body }
                </p>
            </div>

            {/* Fecha del entry */}
            <div className="journal__entry-date-box">
                <span> {`${noteDate.format('dddd')} ${noteDate.format('Do')} `  /*Día de la semana(e.g Sunday) Ver la doc de moment, Día del mes(e.g 26th) */ } </span>
                <h4>{`${noteDate.format('MMM')} ${noteDate.format('YYYY')}` /* Mes y Año */ }</h4>
                
            </div>

        </div>
    )
}
