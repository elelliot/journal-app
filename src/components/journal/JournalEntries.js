import React from 'react'
import { useSelector } from 'react-redux';
import { JournalEntry } from './JournalEntry';

export const JournalEntries = () => {

    //const entries =[1,2,3,4,5,6,7,8,9];

    //obtenemos las notes del state
    const { notes } = useSelector( state => state.notes );
    //console.log(notes);

    return (
        //Es el contenedor de los entries
        <div className="journal__entries">
            
            
            {//Creamos el componente basado en el numero de entries
                // entries.map( value => (
                //     <JournalEntry key={ value } />

                // ))

                notes.map( note => (
                    <JournalEntry
                        key= { note.id }
                        { ...note } //Extraemos todas las propiedades
                    />
                ))
            }


        </div>
    )
}
