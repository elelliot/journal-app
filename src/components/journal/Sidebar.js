import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { startLogout } from '../../actions/auth';
import { startNewNote } from '../../actions/notes';
import { Toast } from '../../alerts/alert';
import { JournalEntries } from './JournalEntries'

export const Sidebar = () => {

    //Obtenemos el username
    const {name} = useSelector( state => state.auth );

    //Manejando el Logout
    const dispatch = useDispatch();
    const handleLogout = () => {
        dispatch( startLogout() );
        Toast.fire({
            iconColor: '#41D70C',
            icon: 'success',
            title: 'Byeee ^u^/'
          })
    }


    //
    const handleAddNew = () => {
        dispatch( startNewNote() );
    }

    return (
        <aside className="journal__sidebar">
            
            <div className="journal__sidebar-navbar">
                <h3 className="mt-5">
                    {/* Icono de lunita de sidebar */}
                    <i className="far fa-moon"></i>
                    <span> {name} </span>
                </h3>

                <button 
                    className="btn"
                    onClick={ handleLogout }
                >
                    Logout
                </button>
            </div>

            <div 
                className="journal__new-entry pointer"
                onClick={ handleAddNew }
                >
                
                {/* √çcono del calendario con signo de + */}
                <i className="far fa-calendar-plus fa-5x"></i>
                <p className="mt-5">
                    New Entry
                </p>
            </div>
            
            <JournalEntries />


        </aside>
    )
}
