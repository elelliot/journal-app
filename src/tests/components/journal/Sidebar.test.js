import { mount } from "enzyme";
import { Provider } from "react-redux";

import configureStore from 'redux-mock-store'; //ES6 modules
import thunk from 'redux-thunk';

import '@testing-library/jest-dom';

import { Sidebar } from "../../../components/journal/Sidebar";
import { startLogout } from "../../../actions/auth";
import { startNewNote } from "../../../actions/notes";


jest.mock('../../../actions/auth', ()=> ({
    startLogout: jest.fn()
}));

jest.mock('../../../actions/notes', ()=> ({
    startNewNote: jest.fn()
}))


const middlewares = [thunk];
const mockStore = configureStore(middlewares);

//Le asignamos el estado que tendría el RegisterScreen, o al menos lo escencial
const initState = {
    auth: {
        uid:'dJ6s8kRzbmay0Nie4QEwl0u7mlB2',
        name: null
    },
    ui:{
        loading: false,
        msgError: null
    },
    notes:{
        notes:[//puede ser datos falsos pero whatever
            {
                id:'4baZszlzfjfKoWQqBu3e',
                body:'Hola 2',
                date:1643968823092,
                title:'Probando Save Note',
                url:'https://gatito2.com/gatito2.jpg'
            },
            {
                id:'TLnQMM2J1vrisWCLhRa8',
                body:'',
                date:1643968948919,
                title:''
            },
            {
                id:'XSTvJIPkXnTW67uFQNoQ',
                body:'',
                date:1643968673039,
                title:''
            },
        ],
        active:null
    }
};



let store = mockStore( initState );
store.dispatch = jest.fn(); //para las async actions


const wrapper = mount(
    <Provider store={ store }>
        <Sidebar />
    </Provider>
);

describe('Pruebas en <Sidebar />', () => {



    test('Debe mostrarse correctamente', () => {
        //snapshot
        expect( wrapper ).toMatchSnapshot();
    });


    test('Debe llamar el startLogout action', () => {

        //Damos Click al logout (se supone que es el único por eso puse el elemento)
        wrapper.find('button').prop('onClick')();
        expect( startLogout ).toHaveBeenCalled();

    });


    test('Debe llamar el startNewNote action', () => {
        //Que se dispare la acción
        wrapper.find('.journal__new-entry').simulate('click');
        expect( startNewNote ).toHaveBeenCalled();

    })

})