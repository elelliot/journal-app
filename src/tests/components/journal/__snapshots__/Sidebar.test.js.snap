// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <Sidebar /> Debe mostrarse correctamente 1`] = `
<aside
  className="journal__sidebar"
>
  <div
    className="journal__sidebar-navbar"
  >
    <h3
      className="mt-5"
    >
      <i
        className="far fa-moon"
      />
      <span>
         
         
      </span>
    </h3>
    <button
      className="btn"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
  <div
    className="journal__new-entry pointer"
    onClick={[Function]}
  >
    <i
      className="far fa-calendar-plus fa-5x"
    />
    <p
      className="mt-5"
    >
      New Entry
    </p>
  </div>
  <div
    className="journal__entries"
  >
    <div
      className="journal__entry pointer animate__animated animate__fadeIn animate__faster"
      onClick={[Function]}
    >
      <div
        className="journal__entry-picture"
        style={
          Object {
            "backgroundImage": "url(https://gatito2.com/gatito2.jpg)",
            "backgroundSize": "cover",
          }
        }
      />
      <div
        className="journal__entry-body"
      >
        <p
          className="journal__entry-title"
        >
          Probando Save Note
        </p>
        <p
          className="journal__entry-content"
        >
          Hola 2
        </p>
      </div>
      <div
        className="journal__entry-date-box"
      >
        <span>
           
          Friday 4th 
           
        </span>
        <h4>
          Feb 2022
        </h4>
      </div>
    </div>
    <div
      className="journal__entry pointer animate__animated animate__fadeIn animate__faster"
      onClick={[Function]}
    >
      <div
        className="journal__entry-body"
      >
        <p
          className="journal__entry-title"
        />
        <p
          className="journal__entry-content"
        />
      </div>
      <div
        className="journal__entry-date-box"
      >
        <span>
           
          Friday 4th 
           
        </span>
        <h4>
          Feb 2022
        </h4>
      </div>
    </div>
    <div
      className="journal__entry pointer animate__animated animate__fadeIn animate__faster"
      onClick={[Function]}
    >
      <div
        className="journal__entry-body"
      >
        <p
          className="journal__entry-title"
        />
        <p
          className="journal__entry-content"
        />
      </div>
      <div
        className="journal__entry-date-box"
      >
        <span>
           
          Friday 4th 
           
        </span>
        <h4>
          Feb 2022
        </h4>
      </div>
    </div>
  </div>
</aside>
`;
